mathics/core/expression.py

```diff
@@ -1165,6 +1250,132 @@ def sameQ(self, other: BaseElement) -> bool:
             for element, oelement in zip(self._elements, other.get_elements())
         )
 
+    def sequences(self):
+        cache = self._cache
+        if cache:
+            seq = cache.sequences
+            if seq is not None:
+                return seq
+
+        return self._rebuild_cache().sequences
+
+    def set_head(self, head):
```

Comment:
This is used only in 3 places:
- algorithms/parts.py
- builtin/structure.py
- core/subexpression.py

Reply:
This and the other comment are all got, but I need to stop working on this. 

@TiagoCavalcante @mmatera I am happy to leave open this PR if someone wants to move it forward. Or we could comment and continue later. 

Discuss and decide.

Reply:
@mmatera do you know whether we really need to invalidate the cache? If that isn't necessary, it would be just set a property to another value, so it could be inlined, otherwise, I think it is better to keep it where it is by now.

Reply:
@TiagoCavalcante @mmatera please stop with the performance improvement suggestions for evaluation. Or make note of that somewhere and we can come back to it. Thanks.

Reply:
@rocky that isn't about performance. My thoughts [rewritten]: if that function does 2 things, then keep it where it is, if the function does 1 thing, inline it in the 3 places it is called.

Reply:
@TiagoCavalcante ok - sorry. 

But a couple of things here. As we've seen too many times in the past if something seems a little bit more complicated, there probably is a reason for this and best to assume that complications were put there for a reason. 

Sadly, these reasons haven't been spelled out so we can't tell if the reason still holds or had a fallacy in it. But to start out we should assume it is true. 

Second, here in the refactor part I am concerned about clarity.  And a-priori I don't see how inlining this is going to clarify something. Adding a docstring to this code would though. But that would mean understanding something about the code more than just where is it used. 

And in general there has been in my opinion too much discussion about things at the level of  inlining pieces of code or the order of if statements versus what is it we are trying to accomplish in the bigger picture. And this keeps us from moving forward towards understanding how to do what is necessary clearly, effectively (and efficiently).

Reply:
@rocky ok, I think it is better to keep it as is then. From my part, we can merge now.